### ISSUE TITLE
Running multiple terminals is filling video memory

### ISSUE THREAD

**Issue by jorisgio** [jwilm/alacritty#864]
I have a low end nividia GT730 with proprietary driver. The graphic card has 2GB of ram. After a day or so, when i have a dozen alacritty opened, the memory usage reported by nividia-smi is around 1.9G, and my whole X11 environment becomes sluggish.
 
Which operating system does the issue occur on?
linux (debian stretch)

If on linux, are you using X11 or Wayland?
X11

version:
0.1.0 (5ac42bb13b)

**Comment by jwilm** [comment]
@jorisgio are you dealing with a lot of non-ascii text?

**Comment by jorisgio** [comment]
I'm running with a few neovim. I guess you are thinking about glyphe cache ? I'm mostly displaying code inside, with FiraCode font, but quite large font size on a 4k screen.

**Comment by jwilm** [comment]
yeah, the only thing that could potentially leak video memory are the textures. I'll look into adding a way to debug this easily and get back to you.

**Comment by grinapo** [comment]
One instance uses 10MB here. It is an interesting effect when I open plenty of them like
```for i in `seq 1 100`; do (./target/debug/alacritty -e /tmp/sleep30.sh &) ; done```
Debug mode doesn't say anything but the windows get stuck for a few minutes without a process around, then just close. In release mode I get a few
```
Event loop exitting due to error: Input/output error (os error 5) [src/event_loop.rs:401]
```
and the windows linger around with their process already gone for a few dozen seconds.

**Comment by jorisgio** [comment]
It's probably the same issue i had indeed, not a leak but high overhead. I tend to use around ~10/15 terminals and with a 4K screen it tend to make my desktop freeze after only a few terminals opened.
I stopped using alacritty and back to urxvt for this reason.

**Comment by jwilm** [comment]
One approach to resolving this problem might be something similar to urxvtd where all of the running alacritty instances can share certain gl allocations like glyph atlases.

**Comment by kchibisov** [comment]
`3840 * 2160 * 4 / 2 ** 20` = 31Mb. Assuming double buffering 62Mb. If it's fullscreen the buffer count can go even higher (all in driver, can not do much about it). Then I few MBs for our stuff.

Like the baseline for buffer is already relatively high, and the only thing we can improve is to share textures between multiple windows, but it won't improve single window and the baseline.

**Comment by paboum** [comment]
> `3840 * 2160 * 4 / 2 ** 20` = 31Mb. Assuming double buffering 62Mb. If it's fullscreen the buffer count can go even higher (all in driver, can not do much about it). Then I few MBs for our stuff.

I don't care. It's a terminal emulator, needs ca. 108x320 bytes, or maybe 4-byte words for UTF characters. Maybe a byte more per character if I enabled colors. There is literally no benefit of storing each uncompressed pixel in VRAM until someone wants to play movies in libcaca. Seems like a flawed design, so I'm no longer using this.

**Comment by kchibisov** [comment]
I'd say that every terminal in existing as well as pretty much any other program (unless you're talking about server side rendering into compositor buffer directly) stores the buffer in a requested pixel format. That simply not how things work, and foot stores the same amount on the CPU side for drawing, because that's just how it works.

You can not compress the buffer itself(I'm not saying textures, I'm talking about where the pixels go). I'd assume you surely understand how to render window into the screen and what implications there're.

Also, for single cell you'd need in modern terminal around 24bytes on 64-bit system, foot uses around that amount nowadays iirc(maybe a bit less, depending on the features it has), but byte per color is an interesting way to compress rgb.

**Comment by kchibisov** [comment]
I'd add a note that it consumes zero vram since it's sw rendering and being copied into compositor's buffer which is in vram, while apps that use gpu allocate their buffer in vram directly, so it's more efficient in the end.

**Comment by grinapo** [comment]
(@kchibisov I am humbled by your patience and empathy against a very rude commenter.)

**Comment by paboum** [comment]
> efficient in the end

You mean it renders in like 0.5 ms instead of 0.75 ms? Wow, go get them! I prefer my VRAM being spent on LLMs and SD models instead.

![Image](https://github.com/user-attachments/assets/2640aef0-9e78-42ca-bf7a-0f16953f8a5f)

**Comment by grinapo** [comment]
@paboum why do you write messages here when you already told us all that you will not use alacritty, it is not good for you, you will use something else, you think the goals of alacritty are not aligning with your goals? It would be best that since you already told everyone that you dislike this project that you would go away for good.

Really, nobody forces you to use it, and now that you realised that it does things you don't need and uses resources that you do need elsewhere you really should stop using it and stop caring about it. I definitely don't need to read repeatedly how you dislike this project. This issue is about _excessive_ use, but your problem is not that, your problem is that _normal_ use is too much for you. This issue is __not__ about that, and will not help that. Please, stop repeating that you don't like that. Thank you!

**Comment by kchibisov** [comment]
No, it uses less battery and may also result in lower response times since you copy less and have less rendering operations at the end of the day. That's the whole reason video hw accelerated video playback is so energy efficient, etc, etc. At the end of the day app content is ending up on the GPU anyway, so if you start from it you have one fewer copy or zero copy at all (direct scanout, can be used).

I'd note that I only reply to correct false assumptions, I don't really care if the person uses it at the end of the day or not, like I'm not dining that we use vram, because we use OpenGL and that on 4k because buffers are generally large we use more, even though, we use texture atlases, etc, etc. Ofc software rendering won't use vram at all, since it's software, will just use ram for those buffers.

**Comment by paboum** [comment]
Why u keep asking questions and making false statements? Just close as "not planned".

Terminal emulators existed in, say, 2001, when the average CPU was like Pentium 200 MMX. How much CPU could a terminal emulator consume back then to be even noticable for a modern hardware? Don't answer, this is a rhetorical question. Intelligent readers will already understand and no amount of explanation will help the others. Or another one: which one is more expensive and/or less available, RAM or VRAM? Or maybe the best one yet: why does foot fit in less RAM than VRAM alacritty needs? But hey, don't fret, I used kitty before: it uses BOTH like 189 MB of RAM AND 76 MB of VRAM. I really considered shifting from kitty to alacritty a good choice, still do.

![Image](https://github.com/user-attachments/assets/01bb5c6c-7bd8-4b75-b850-f6c863bfa544)

**Comment by paboum** [comment]
This one I can't even believe: https://codeberg.org/dnkl/foot/src/branch/master/doc/benchmark.md
foot beats alacritty even in its' own benchmark?
It really seems like some educated guy observed the same thing the op here did and decided to show how it's supposed to be done. Are you guys genzies with no computer science diploma? Please, don't answer. I'm unsubscribing from notifications on this thread.

**Comment by jwilm** [comment]
@paboum while Alacritty welcomes critical feedback of the project, we certainly do not welcome personal attacks against our community or our maintainers. If you choose to continue such combative interactions on our issue tracker, we'll be blocking you from participating indefinitely. This is your only warning.

To everyone else in this thread, thank you for maintaining a high degree of of decorum. There's certainly some optimizations to be made with how Alacritty leverages VRAM. Let's please focus on the issue at hand.

**Comment by paboum** [comment]
> This is your only warning.

You seem not to understand how open communities work. Whenever I wish to report an issue, it will be reported from any working GH account. I can also spend thousands of dollars just to donate foot or make fun of your incompetence on youtube. Especially when you try to fight criticism. This is literally how poor projects die, should you choose not to fix it.

**Comment by chrisduerr** [comment]
I'd recommend anyone who thinks they're affected by this issue to read up on how VRAM works first. While I don't mind laughing at your incompetence, I recommend not making a public mockery of yourself. Unfortunately most people are far too kind with trolls.
